File: src/main/java/gargoyle/xenox/game/sys/HiScore.java
package gargoyle.xenox.game.sys;

import gargoyle.xenox.util.config.Config;
import gargoyle.xenox.util.log.Log;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.prefs.BackingStoreException;
import java.util.prefs.Preferences;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public final class HiScore extends Config {
    public static final int LINES = 10;
    private static final Pattern DIGITS = Pattern.compile("[0-9]+");
    private static final String EMPTY = "";
    private static final String PARAM_NAME = "name";
    private static final String PARAM_SCORE = "score";
    private static final long serialVersionUID = -8306160552821206611L;

    public Record get(int index) {
        String key = String.valueOf(index);
        return new Record(node(key).get(PARAM_NAME, EMPTY), node(key).getInt(PARAM_SCORE, 0));
    }

    public List<Record> getRecords() {
        int size = size();
        List<Record> records = new ArrayList<>(size);
        for (int i = 0; i < size; i++) {
            records.add(get(i));
        }
        return Collections.unmodifiableList(records);
    }

    public String getUserName() {
        return get(PARAM_NAME, EMPTY);
    }

    public void setUserName(String username) {
        put(PARAM_NAME, username);
    }

    private void put(int index, String name, int score) {
        Preferences node = node(String.valueOf(index));
        node.put(PARAM_NAME, name);
        node.putInt(PARAM_SCORE, score);
    }

    public void score(String name, int score) {
        put(size(), name, score);
        List<Record> table = IntStream.range(0, size()).mapToObj(this::get).sorted().collect(Collectors.toList());
        IntStream.range(0, Math.min(LINES, table.size())).forEach(index -> put(index, table.get(index).getName(), table.get(index).getScore()));
    }

    public int size() {
        int size = 0;
        try {
            for (String key : childrenNames()) {
                if (DIGITS.matcher(key).matches()) {
                    size = Math.min(LINES, Integer.parseInt(key) + 1);
                    String name = node(key).get(PARAM_NAME, EMPTY);
                    if (name != null && name.isEmpty()) {
                        break;
                    }
                }
            }
        } catch (BackingStoreException e) {
            Log.error(e.getLocalizedMessage(), e);
        }
        return size;
    }

    public static final class Record implements Serializable, Comparable<Record> {
        private final String name;
        private final int score;

        private Record(String name, int score) {
            this.name = name;
            this.score = score;
        }

        @Override
        public int compareTo(Record o) {
            return Integer.compare(o.score, score);
        }

        public String getName() {
            return name;
        }

        public int getScore() {
            return score;
        }

        @Override
        public int hashCode() {
            int prime = 31;
            int result = 1;
            result = prime * result + (name == null ? 0 : name.hashCode());
            return result;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (obj == null || getClass() != obj.getClass()) {
                return false;
            }
            Record other = (Record) obj;
            return name == null ? other.name == null : name.equals(other.name);
        }
    }
}

File: src/main/kotlin/gargoyle/xenox/game/sys/HiScore.kt
package gargoyle.xenox.game.sys

import gargoyle.xenox.util.config.Config
import gargoyle.xenox.util.log.Log
import java.io.Serializable
import java.util.Collections
import java.util.prefs.BackingStoreException
import java.util.regex.Pattern
import java.util.stream.Collectors
import java.util.stream.IntStream

class HiScore : Config() {
    operator fun get(index: Int): Record {
        val key = index.toString()
        return Record(node(key)[PARAM_NAME, EMPTY], node(key).getInt(PARAM_SCORE, 0))
    }

    val records: List<Record>
        get() {
            val size = size()
            val records: MutableList<Record> = ArrayList(size)
            for (i in 0 until size) {
                records.add(get(i))
            }
            return Collections.unmodifiableList(records)
        }
    var userName: String?
        get() = get(PARAM_NAME, EMPTY)
        set(username) {
            put(PARAM_NAME, username!!)
        }

    private fun put(index: Int, name: String?, score: Int) {
        val node = node(index.toString())
        node.put(PARAM_NAME, name)
        node.putInt(PARAM_SCORE, score)
    }

    fun score(name: String?, score: Int) {
        put(size(), name, score)
        val table = IntStream.range(0, size()).mapToObj { index: Int -> this[index] }
            .sorted().collect(Collectors.toList())
        IntStream.range(0, LINES.coerceAtMost(table.size))
            .forEach { index: Int -> put(index, table[index].name, table[index].score) }
    }

    fun size(): Int {
        var size = 0
        try {
            for (key in childrenNames()) {
                if (DIGITS.matcher(key).matches()) {
                    size = LINES.coerceAtMost(key.toInt() + 1)
                    val name = node(key)[PARAM_NAME, EMPTY]
                    if (name != null && name.isEmpty()) {
                        break
                    }
                }
            }
        } catch (e: BackingStoreException) {
            Log.error(e.localizedMessage, e)
        }
        return size
    }

    class Record(name: String, score: Int) : Serializable, Comparable<Record> {
        val name: String?
        val score: Int

        init {
            this.name = name
            this.score = score
        }

        override fun compareTo(other: Record): Int {
            return other.score.compareTo(score)
        }

        override fun hashCode(): Int {
            val prime = 31
            var result = 1
            result = prime * result + (name?.hashCode() ?: 0)
            return result
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }
            if (other == null || javaClass != other.javaClass) {
                return false
            }
            return if (name == null) (other as Record).name == null else name == (other as Record).name
        }
    }

    companion object {
        const val LINES = 10
        private val DIGITS = Pattern.compile("[0-9]+")
        private const val EMPTY = ""
        private const val PARAM_NAME = "name"
        private const val PARAM_SCORE = "score"
        private const val serialVersionUID = -8306160552821206611L
    }
}
